pages/XX-ai-methods/reproducible_prompts/Reproducible_Prompts.qmd

```diff
@@ -0,0 +1,299 @@
+Ollama is a framework that allows for the installation of many LLM models
+locally. We need to choose a starting model and then install it. A popular
+option is the llama model. While not open source it has fairly permissive
+[terms](https://www.llama.com/llama3/license/) of use.
+
+::: {.callout-note}
+We will be running a console command here. We will be using the integrated
+jupyter magic command `!` which will send our code to the system shell. Make sure the ollama server is running, or this will give an error.
+
+:::
+
+```{python}
+# Meta's Llama 3.2 goes small with 1B and 3B models. 3B is default
+# Licensed under the llama Community Licence Agreement 
+!ollama pull llama3.2
+```
+
+We can check that the model was successfully installed with the
+`ollama list` command.
+
+```{python}
+!ollama list
+```
+
+# Jupyter AI **🪄Magic🪄** 
+## Now let's play
+- We already loaded the ai_magicks extension to use it in our notebook
+- We'll also just set the AiMagics max_history to 0, so it wont remember any past prompt information in the current prompt
+  - This is to make things a little more repeatable
+  - ::: {.callout-warning}
+    The docs suggests using `%ai reset` to clear the chat history, DON'T DO THIS. It also clears any parameters you have set, and seems to actually hurt reproducibility.
+    
+    :::
+
+```{python}
+%config AiMagics.max_history = 0
+```
+
+Now we can ask llama3 to introduce itself. There are a few important parts of
+the code we must include:
+1.  `%%ai` Designates the code cell as an ai cell. We can then treat it as a 
+    chat window.
+2.  `ollama:llama3.2` tells jupyter that the cell well use the ollama interface.
+    Additionally we have to specify the actual LLM because ollama can any model
+    you have installed.
+3.  `-m {}` allows you to provide additionall arguments to your LLM provider. In
+    our case we will be providing the default ollama port on local host
+    `"base_url":"http://localhost:11434"` so jupyter knows where to look for
+    ollama.
+
+```{python}
+%%ai ollama:llama3.2 -m {"base_url":"http://localhost:11434"}
+Introduce yourself
+```
+
+## Now let's try incoporating a python object
+We can incoporate a python object using the `{}` syntax inside our prompt. First
+Lets make a list in python.
+
+```{python}
+random_words = ["cat", "dog", "bird", "monkey", "elephant", "deer", "wolf", "bison", "elk"]  
+print(random_words)
+```
+
+Now we can embed this list in our prompt by calling the variable name inside of
+`{}`
+
+```{python}
+#| scrolled: true
+%%ai ollama:llama3.2 -m {"base_url":"http://localhost:11434"}
+Create a bullet list from the text of {random_words} in random order
+```
+
+We can set the output format using the `-f` flag. Lets specify we would like
+our output in json.
+
+```{python}
+%%ai ollama:llama3.2 -f json -m {"base_url":"http://localhost:11434"}
+Create a list from the text of {random_words} in random order
+```
+
+## Lets keep making things more reproducible
```

Comment:
I like the style of these headings, very conversational. That said, keep in mind that they are used to generate the "On this page" navigation. I think it might be worth re-working some of the top-level headers to make the page more navigable. I don't have any kind of style guide here yet -- just something I've been thinking about with my own writing.